java/src/main/java/com/google/appengine/tools/mapreduce/outputs/GoogleCloudStorageFileOutputWriter.java

```diff
@@ -32,88 +30,123 @@
+        client = StorageOptions.newBuilder()
+          .setCredentials(this.options.getServiceAccountCredentials().get())
+          .setProjectId(this.options.getProjectId())
+          .build().getService();
+      } else {
+        client = StorageOptions.getDefaultInstance().getService();
+      }
+    }
+    return client;
+  }
+
   @Override
   public void beginShard() throws IOException {
-    GcsFileOptions fileOptions = new GcsFileOptions.Builder().mimeType(mimeType).build();
-    GcsFilename dest = new GcsFilename(file.getBucketName(), file.getObjectName() + "~");
-    channel = GCS_SERVICE.createOrReplace(dest, fileOptions);
+    Blob shardBlob = getClient().create(getWorkingLocation());
+    shardBlobId = BlobId.of(shardBlob.getBucket(), shardBlob.getName());
     sliceChannel = null;
     toDelete.clear();
   }
 
   @Override
   public void beginSlice() throws IOException {
     cleanup();
-    if (supportSliceRetries) {
-      if (sliceChannel != null) {
-        copy(sliceChannel.getFilename(), channel);
-        toDelete.add(sliceChannel.getFilename().getObjectName());
+    if (options.getSupportSliceRetries()) {
+      if (sliceBlobId != null) {
+        //append latest version of previous slice's file to shard's file
+        // q: why not do this in endSlice??
+        // q: why are we doing this as part of every slice? why not just do a big compose of all slices
+        // into shard at endShard? a: bc may
+        BlobId lastSliceBlobId = BlobId.of(sliceBlobId.getBucket(), sliceBlobId.getName());
+
+        //q: race condition here? what if this append() doesn't really see the 'latest' copy of blob?
+        append(lastSliceBlobId, shardBlobId);
+        toDelete.add(lastSliceBlobId);
       }
-      GcsFileOptions opt = new GcsFileOptions.Builder().mimeType(mimeType).build();
-      String name = file.getObjectName() + "~" + Math.abs(RND.nextLong());
-      sliceChannel = GCS_SERVICE.createOrReplace(new GcsFilename(file.getBucketName(), name), opt);
+      Blob sliceBlob = getClient().create(getSliceWorkingLocation());
+      sliceBlobId = BlobId.of(sliceBlob.getBucket(), sliceBlob.getName());
+      sliceChannel = sliceBlob.writer();
     } else {
-      sliceChannel = channel;
+      //if won't retry slices, can just write straight into shard's blob
+      sliceChannel = getClient().get(shardBlobId).writer();
     }
+    sliceChannel.setChunkSize(DEFAULT_IO_BUFFER_SIZE);
   }
 
-  private static void copy(GcsFilename from, GcsOutputChannel toChannel) throws IOException {
+
+  void append(BlobId src, BlobId dest) throws IOException {
     ByteBuffer buffer = ByteBuffer.allocate(MapReduceConstants.DEFAULT_IO_BUFFER_SIZE);
-    try (GcsInputChannel fromChannel = GCS_SERVICE.openReadChannel(from, 0)) {
-      while (fromChannel.read(buffer) >= 0) {
+    WriteChannel destChannel =
+      getClient().writer(BlobInfo.newBuilder(dest.getBucket(), dest.getName()).build());
```

Comment:
```suggestion
      getClient().writer(BlobInfo.newBuilder(dest).build());
```
There are many methods where you can pass the `blobId` directly, less error prone